libvast/vast/transform_step.hpp

```diff
@@ -0,0 +1,74 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/table_slice.hpp"
+
+#include <caf/expected.hpp>
+
+#if VAST_ENABLE_ARROW > 0
+#  include <arrow/record_batch.h>
+#endif
+
+namespace vast {
+
+/// An individual transform step. This is mainly used in the plugin API,
+/// later code deals with a complete `transform`.
+class transform_step {
+public:
+  virtual ~transform_step() = default;
+
+  /// Apply the transformation step to the passed table slice,
+  /// if neccessary converting the data to a format the step implementation can
+  /// natively handle.
+  [[nodiscard]] caf::expected<table_slice> apply(table_slice&&) const;
+};
+
+using transform_step_ptr = std::unique_ptr<transform_step>;
+
+// A transform step that operates on a generic table slice.
+// While the implementation is free to look at the encoding of
+// the slice and dispatch to specialized functions, this overload
+// requires the result of the step to be serialized back into
+// a finished table slice, while the variants below can pass around
+// intermediate state between the steps.
+class generic_transform_step : public virtual transform_step {
+public:
+  [[nodiscard]] virtual caf::expected<table_slice>
+  operator()(table_slice&&) const = 0;
+};
+
+// TODO: This will become useful once we implement filtering transforms,
+//       where msgpack encoding allows for higher performance.
+// class msgpack_transform_step : public virtual transform_step {
+//   [...]
+// };
+
+#if VAST_ENABLE_ARROW > 0
```

Comment:
This can be rewritten as `#if VAST_ENABLE_ARROW`

Reply:
When doing this please grep for `VAST_ENABLE_ARROW > 0` generally. There are a few more places.